browser/ui/tabs/shared_pinned_tab_service.cc

```diff
@@ -0,0 +1,731 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/tabs/shared_pinned_tab_service.h"
+
+#include <utility>
+
+#include "brave/browser/ui/tabs/features.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "chrome/browser/ui/browser_tabrestore.h"
+#include "chrome/browser/ui/browser_tabstrip.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+#define LOCK_REENTRANCE(tab_strip_model)                                  \
+  DCHECK(!change_source_model_) << "Already locked";                      \
+  base::AutoReset<raw_ptr<TabStripModel>> resetter(&change_source_model_, \
+                                                   raw_ptr(tab_strip_model))
+
+namespace {
+
+////////////////////////////////////////////////////////////////////////////////
+// SharedContentsData is a WebContentsUserData attached to pinned tab's web
+// contents that could be movable between multiple windows.
+//
+class SharedContentsData
+    : public content::WebContentsUserData<SharedContentsData> {
+ public:
+  ~SharedContentsData() override = default;
```

Comment:
From other comment you said,

> AFAIK, they are not generated implicitely when configure copy ctors explicitely.

If you prefer deleting only when it could be implicilty generated, I think we could omit deleting copy & assign ctors here, because they're not deleted from the base class. What do you think?

Reply:
If you want to allowing copy the instance of `SharedContentsData`, you don't need to delete explicitely.
but I think you don't want?

Reply:
As base class deleted the copy constructor, we can't copy SharedContentsData even without deleting explicitly from child class.

```
#include <iostream>
using namespace std;

class A {
 public:
    A() = default;
	A(const A&) = delete;
	A& operator=(const A&) = delete;
};

class B : public A {
 public:
	B() = default;
};

int main() {
	// your code goes here
	B b1;
	B b2(b1);  // error: use of deleted function ‘B::B(const B&)’
	return 0;
}
```

Reply:
Ugh, because of this complicated generation rule, I prefer the rule of 5 on the base class 😄

Reply:
Hmm, all base classes of `SharedContentsData` doesn't delete them?

Reply:
Oops, I mistook SupportsUserData for SupportsUserData::Data 😭

Reply:
I assumed that :)